libsolidity/analysis/ContractLevelChecker.cpp

```diff
@@ -133,52 +163,217 @@ void ContractLevelChecker::findDuplicateDefinitions(map<string, vector<T>> const
 	}
 }
 
+vector<ContractDefinition const*> resolveBaseContracts(ContractDefinition const& _contract)
+{
+	vector<ContractDefinition const*> resolvedContracts{_contract.baseContracts().size(), nullptr};
+
+	transform(
+		_contract.baseContracts().cbegin(),
+		_contract.baseContracts().cend(),
+		resolvedContracts.begin(),
+		[] (ASTPointer<InheritanceSpecifier> const& _specifier)
+		{
+			Declaration const* baseDecl =
+				_specifier->name().annotation().referencedDeclaration;
+			return dynamic_cast<ContractDefinition const*>(baseDecl);
+		}
+	);
+
+	return resolvedContracts;
+}
+
+ContractLevelChecker::FunctionInfo ContractLevelChecker::findAndCheckOverride(
+	ContractDefinition const& _derived,
+	FunctionDefinition const& _function,
+	ContractDefinition const& _base
+)
+{
+	// Look for possible override in all functions of _base
+	for (FunctionDefinition const* baseFunc: _base.definedFunctions())
+	{
+		if (!hasEqualNameAndParameters(_function, *baseFunc))
+			continue;
+
+		if (!checkFunctionOverride(_function, *baseFunc))
+			return {};
+
+		return {baseFunc, &_base};
+	}
+
+	// If we found no matching override, check the bases of base
+	for (auto const c: resolveBaseContracts(_base))
+	{
+		auto const& baseBase = findAndCheckOverride(_derived, _function, *c);
+
+		if (baseBase.funcDef)
+			return baseBase;
+	}
+
+	return {};
+}
+
 void ContractLevelChecker::checkIllegalOverrides(ContractDefinition const& _contract)
 {
+	for (FunctionDefinition const* function: _contract.definedFunctions())
+	{
+		vector<FunctionInfo> baseFuncs;
+
+		if (function->isConstructor())
+			continue;
+
+		// Collect all functions that are overriden by this one
+		for (auto const c: resolveBaseContracts(_contract))
+		{
+			auto result = findAndCheckOverride(_contract, *function, *c);
+
+			if (result.funcDef)
+			{
+				auto baseResult = find(
+					baseFuncs.cbegin(),
+					baseFuncs.cend(),
+					result
+				);
+
+				if (baseResult == baseFuncs.cend())
+					baseFuncs.emplace_back(result);
+			}
+		}
+
+		// Missing override errors are reported in checkFunctionOverride()
+		if (!function->overrides())
+			continue;
+
+		// Only one override and no list specified is okay
+		if (baseFuncs.size() == 1 && function->overrides()->overrides().empty())
+			continue;
+
+		vector<ContractDefinition const*> specifiedContracts =
+			resolveOverrideList(*function->overrides());
+
+		decltype(specifiedContracts) missingContracts;
+
+		// Validate override list
+		for (FunctionInfo const& overriden: baseFuncs)
+		{
+			auto const result = find(
```

Comment:
Please use `.count()`

Reply:
While you can't use ``.count()``, you can still use ``std::count(...)`` or even better the ranged version of ``boost::count(...)``.

Reply:
Or use ``contains`` from libdevcore/CommonData ;-).

Reply:
or our function `contains()` ..

Reply:
Yeah ok - but I think this whole loop here can still be simplified, but I'll have another look tomorrow.